### PR TITLE
graalvmXX-ee: remove; graalvmXX-ce: remove Ruby/Python/WASM support

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Description of changes

This PR does two things:
- Removes `graalvmXX-ee`: this was last updated in 2020, and it is really out-of-date. Since this also can't be build automatically (because access to the required files is restrict), we can't even be sure if this still builds. So remove it, if someone complains we can revert later.
- Removes support for Python/Ruby/WASM in `graalvmXX-ce`. This is not used anywhere in nixpkgs itself, and also IMO it is not used a lot at all. For example, Ruby support is broken for a long time already (according to our tests), yet nobody bothered to try to fix it. Keeping support for it is just increasing the support burden and increasing the derivation size without many benefits. It is still possible to build it if you want though, since the code necessary to build with support for it is still there. You will need to use a customized `pkgs.graalvmCEPackages.mkGraal` call for it, but it shouldn't be too difficult.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 168816` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>graalvm11</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>HentaiAtHome</li>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>dapl-native</li>
    <li>dbqn-native</li>
    <li>graalvm11-ce</li>
    <li>graalvm17-ce</li>
    <li>jet</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
@ofborg eval babashka clojure-lsp HentaiAtHome

**Comment by thiagokokada** [comment]
@ofborg build babashka clojure-lsp HentaiAtHome

**Comment by thiagokokada** [comment]
CC @tricktron for #166537
Also @glittershark @babaviere @ericdallo @hlolli

**Comment by thiagokokada** [comment]
Before:
```
nix path-info -S /nix/store/sc9vbxw116p9pa0knh0cq7nwzrwq5sya-graalvm11-ce -h
/nix/store/sc9vbxw116p9pa0knh0cq7nwzrwq5sya-graalvm11-ce	  1.9G
```

After:
```
$ nix path-info -S ./result -h
/nix/store/nc8j52ykqbnx7gqi9p0rcq45mp8i9x68-graalvm11-ce-22.0.0.2	  1.5G
```

**Review by superherointj — APPROVED** [review]

**Review by tricktron — COMMENTED** [review]
I really like the cleanup👍🏼.

I remember that the native-image worked on aarch64-linux in my vm in my last pr. I will test that tomorrow and if it still works I guess we can enable that again.

**Comment by thiagokokada** [comment]
> I really like the cleanup👍🏼.
> 
> I remember that the native-image worked on aarch64-linux in my vm in my last pr. I will test that tomorrow and if it still works I guess we can enable that again.

Well, at least when I tried via QEMU emulation mode it didn't work.

The strange thing is that I only get this issue in `graalvm17-ce`. `graalvm11-ce` works fine.

**Comment by glittershark** [comment]
I'm in support of this.

**Comment by thiagokokada** [comment]
@ofborg build babashka clojure-lsp clj-kondo HentaiAtHome

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 168816` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>graalvm11</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>HentaiAtHome</li>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>dapl-native</li>
    <li>dbqn-native</li>
    <li>graalvm11-ce</li>
    <li>graalvm17-ce</li>
    <li>jet</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

**Review by ericdallo — APPROVED** [review]

File: pkgs/development/compilers/graalvm/community-edition/default.nix [hunk 1]
```diff
@@ -32,54 +30,52 @@ in
...
       };
+      aarch64-linux = {
+        # TODO: error with native-image-installable-svm in aarch64-linux:
+        # "Caused by: java.io.IOException: Cannot run program
+        # "/nix/store/00000000000000000000000000000000-gcc-wrapper-9.3.0/bin/gcc" (in
+        # directory"/tmp/SVM-0000000000000000000"): error=0, Failed to exec spawn
+        # helper: pid: 19865, exit value: 1"
+        products = [ "graalvm-ce" ];
+        arch = "linux-aarch64";
+      };
```
Comment (tricktron) [inline pkgs/development/compilers/graalvm/community-edition/default.nix:78]:
```suggestion
      aarch64-linux = {
        inherit products;
        arch = "linux-aarch64";
      };
```

Comment (tricktron) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:78]: @thiagokokada This built in my vm successfully. Can we push these changes and double check with ofborg?

Don't forget to run the update-script to update the sources:smiley:.

**Review by tricktron — CHANGES_REQUESTED** [review]

**Comment by thiagokokada** [comment]
@ofborg build HentaiAtHome

**Comment by thiagokokada** [comment]
@ofborg build graalvm17-ce

**Comment by thiagokokada** [comment]
Seems that `graalvm17-ce` is working fine with Native Image in aarch64-linux, so it seems ready to merge.

Nice that we got to remove most of those TODOs from the code (ok, some of them is not because they're fixed but just because we are not building Ruby/Python/WASM anymore). The code seems in a much better state now.

**Comment by thiagokokada** [comment]
Thanks folks!

**Comment by solson** [comment]
I actively use TruffleRuby and have been in the process of investigating a fix for it. This will make the investigation just slightly harder to do, I guess — the customized call to get it back is like [shown here](https://github.com/NixOS/nixpkgs/pull/168908)?

Regarding the fix, here's one issue my friend and I opened upstream so far: https://github.com/oracle/graal/issues/4478

It's a pretty interesting issue - the Nix builder was running `patchelf` over a bunch of shared libraries that actually contain LLVM bitcode for Sulong (they are Ruby C extensions compiled for TruffleRuby), and the changes `patchelf` made exposed a bug in Sulong's implementation of ELF loading. Hopefully they fix this upstream, but I also experimentally verified that I can build a much-less-broken TruffleRuby locally by making `patchelf` skip those files. There are a few kinks to iron out, and I've been busy, but I may be able to make a nixpkgs PR for this soon.

I had a related general question: is the nixpkgs derivation still only fetching release jars from upstream? I am curious what it would take to make building from source work, because that might give us some more flexibility around avoiding patchelf. I can see how it might be a big effort, though.

By the way, thank you to everyone who's worked on packaging GraalVM for nixpkgs. It's very interesting stuff and I've started successfully experimenting with implementing a Truffle language. :slightly_smiling_face:

**Comment by thiagokokada** [comment]
> I actively use TruffleRuby and have been in the process of investigating a fix for it. This will make the investigation just slightly harder to do, I guess — the customized call to get it back is like [shown here](https://github.com/NixOS/nixpkgs/pull/168908)?

I am not against supporting other languages in our GraalVM setup, however if we do want to do this I would much prefer a composable approach. Like, we could have separate derivations, one for each programming language, and them we could have something like:

```nix
{ graalvm-ce-custom = graalvm17-ce.overrides { extraSVMs = with graalvmCEPackages; [ PythonSVM RubySVM ]; }
```

This way, we can keep maintenance of each separate language in their own individual derivations, instead of having a huge one where everything needs to be working instead.

> Regarding the fix, here's one issue my friend and I opened upstream so far: [oracle/graal#4478](https://github.com/oracle/graal/issues/4478)
> 
> It's a pretty interesting issue - the Nix builder was running `patchelf` over a bunch of shared libraries that actually contain LLVM bitcode for Sulong (they are Ruby C extensions compiled for TruffleRuby), and the changes `patchelf` made exposed a bug in Sulong's implementation of ELF loading. Hopefully they fix this upstream, but I also experimentally verified that I can build a much-less-broken TruffleRuby locally by making `patchelf` skip those files. There are a few kinks to iron out, and I've been busy, but I may be able to make a nixpkgs PR for this soon.

Nice to know that someone is looking for those issues :+1:. While I don't want to ever re-introduce all programming languages in the same package as we used to have in the past, the approach that I proposed above would easily get a +1 for me.

But of course, if you're not up to this you can also just commit the fixes inside the current derivation and leave it behind a the `withRubySvm` flag. This has the issue of not being tested once we update the GraalVM again, but it will work at least for your case.

> I had a related general question: is the nixpkgs derivation still only fetching release jars from upstream? I am curious what it would take to make building from source work, because that might give us some more flexibility around avoiding patchelf. I can see how it might be a big effort, though.

We used to build GraalVM from sources in the past, but it was a huge monumental effort to do each update that eventually it started to get stale, until I removed some time ago: https://github.com/NixOS/nixpkgs/pull/141812.

As you can see by it, it needed a huge amount of patches and it was by far one of the most complex derivations I have ever seen in `nixpkgs`. However, if you want to try it again this can serve as a base.

BTW, you will need a very beefy machine to build this, because AFAIR this failed to build on Hydra because of lack of memory (I think if we set `requiredSystemFeatures = [ "big-parallel" ];` it may have worked, but we never did this). I got it to build successfully with 32GB of RAM, but YMMV.

**Comment by thiagokokada** [comment]
Also, I probably need to eventually create a GraalVM team to maintain the GraalVM packages, and also get some help in writing proper documentation, fixing the issues, etc.

If someone has interested you can speak to me here or in Matrix.

**Comment by solson** [comment]
> I would much prefer a composable approach. Like, we could have separate derivations, one for each programming language

Fully agreed, this would be the ideal outcome for me as well.

> As you can see by it, it needed a huge amount of patches and it was by far one of the most complex derivations I have ever seen in `nixpkgs`.

Oof, yeah, fair enough. I was worried it might be like this, since when I looked into building TruffleRuby from source myself the build system seemed deeply nix-unfriendly.

> Also, I probably need to eventually create a GraalVM team to maintain the GraalVM packages, and also get some help in writing proper documentation, fixing the issues, etc.

The amount of time I can commit to this will be variable, but there are definitely a few issues I would love to help with (right now, primarily TruffleRuby C extensions (as mentioned above) and native-image debug info, which also appears to be broken in an unusual way).

**Pull request merged**